### DISCUSSION TITLE
No live collaboration, server connection lost & realtime websocket error

### DISCUSSION THREAD

**Discussion by ibonelli** [outline/outline#3842] [Self Hosting] [answered]
I have a brand new self-hosted dockerized version of the server which is not being able to work on editing mode. It works to show, it creates new documents, but you can not edit them or add content.

**To Reproduce**
1. Create a new dockerized version of the server
2. Go to any document
3. Click on the content
4. Wait for the document to enter edit mode
5. It doesn't happen, you get the icon "Server connection lost, Edits you make will sync once you’re online"

**Expected behavior**
For the collaboration mode to work. I guess I could disable it, but previously I was running it on Heroku without problems.

**Additional server information**

The server shows the message:

```
outline-outline-1       | {"label":"http","level":"info","message":"  <-- GET /collaboration/document.95dd9287-a5a2-402f-90ce-3fa61987f2e6"}
outline-outline-1       | {"label":"http","level":"info","message":"  --> GET /collaboration/document.95dd9287-a5a2-402f-90ce-3fa61987f2e6 200 12ms 851b"}
outline-outline-1       | {"label":"http","level":"info","message":"  <-- GET /collaboration/document.95dd9287-a5a2-402f-90ce-3fa61987f2e6"}
outline-outline-1       | {"label":"http","level":"info","message":"  --> GET /collaboration/document.95dd9287-a5a2-402f-90ce-3fa61987f2e6 200 12ms 851b"}
```

Tried in two different browsers (details below) and I get the same in the console:

```
hocuspocus-provider.esm.js:1870 WebSocket connection to 'wss://newhandbook.weknowinc.com/collaboration/document.95dd9287-a5a2-402f-90ce-3fa61987f2e6' failed: 
(anonymous) @ hocuspocus-provider.esm.js:1870
createWebSocketConnection @ hocuspocus-provider.esm.js:1868
makeAttempt @ index.js:139
onError @ index.js:100
websocket.js:87 WebSocket connection to 'wss://newhandbook.weknowinc.com/realtime/?EIO=4&transport=websocket' failed: 
```

**Screenshots**

This is what I see in the browser console:

![Screenshot_20220722_160519](https://user-images.githubusercontent.com/7817148/180511413-12b472d7-b5a3-4cac-8b9f-9248f2858cd3.png)

And at the same time the server shows the log:

![Screenshot_20220722_161117](https://user-images.githubusercontent.com/7817148/180511652-b1a64d21-2844-4cb6-9329-51e515237300.png)

**Outline (please complete the following information):**
- Install: Self hosted as a Dockerized version following the github shared recipe.
- Version: Running from image outlinewiki/outline:0.64.3

**Desktop (please complete the following information):**
 - OS: Linux
 - Browser Chrome (v 103) & Firefox (v 91.7)

**Server environment.**
- WEBSOCKET=true
- ENVIRONMENT=development
- DEBUG=http,collaboration

**Comment by ibonelli** [comment]
Forgot to mention that server seems to start fine:

```
outline-outline-1       | {"label":"lifecycle","level":"info","message":"Starting collaboration service"}
outline-outline-1       | {"label":"lifecycle","level":"info","message":"Starting websockets service"}
outline-outline-1       | {"label":"lifecycle","level":"info","message":"Starting worker service"}
outline-outline-1       | {"label":"lifecycle","level":"info","message":"Starting web service"}
outline-outline-1       | {"label":"lifecycle","level":"info","message":"Listening on http://localhost:3000"}
```

And aside from the editing problem, the other server interactions are succesful. Plus as the screenshot shows there is a **collaboration** communication. I see no server mention of a request to **realtime** in the log. The problem only shows in the browser console.

**Comment by ibonelli** [comment]
After reading https://github.com/outline/outline/discussions/3637 I tried rebuilding it from scratch, but got exactly the same result.

**Comment by 1itt1eB0y** [comment, marked as answer]
Did you proxy your outline? If your outline is proxied, you need to enable the websocket function
![图片](https://user-images.githubusercontent.com/29831189/180596909-0165bdfa-ea6c-4f96-af9f-adc6ea0b47bf.png)

**Reply by ibonelli** [reply]
Hi there! That's for that, it fixed the problem. Tweaked a bit the proxy configuration and it now works.

Just for the benefit of anyone else having this issue, I'll share a bit more detail. I'm using the [docker recipe shared by the project's documentation](https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/docker-7pfeLP5a8t#h-docker-compose) which I had to tweak a bit in terms of the included proxy. Even though the [outline repo doesn't have the proxy configuration](https://github.com/outline/outline/blob/main/docker-compose.yml), I found the [project's documentation](https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/docker-7pfeLP5a8t) a bit more complete in terms of what's required to have a docker installation working. It includes the use of [SteveLTN https-portal proxy](https://github.com/SteveLTN/https-portal) which has a couple of configuration you should add to your docker-compose.yml file:

- [DOMAINS](https://github.com/SteveLTN/https-portal#quick-start): The shared example uses a wordpress configuration. For outline dockerized solution you can use ```DOMAINS: 'outline.example.com -> http://outline:3000'```, this refers to the outline container and works as expected.
- [STAGE](https://github.com/SteveLTN/https-portal#quick-start): You have 3 configurations (local, stage & production). You only get a trusted SSL certificate with ```STAGE: production```.
- [WEBSOCKET](https://github.com/SteveLTN/https-portal#websocket): Thanks to the help shared in this discussion  I realized I was missing the ```WEBSOCKET: true``` configuration.

With these 3 added lines in the ```environment``` section things worked as expected:

```
  https-portal:
    image: steveltn/https-portal
    ports:
      - '80:80'
      - '443:443'
    links:
      - outline
    restart: always
    volumes:
      - https-portal-data:/var/lib/https-portal
    healthcheck:
      test: ["CMD", "service", "nginx", "status"]
      interval: 30s
      timeout: 20s
      retries: 3
    environment:
      DOMAINS: 'outline.example.com -> http://outline:3000'
      STAGE: production
      WEBSOCKET: true
```

Just as a personal note, I used the [traefik proxy](https://traefik.io/traefik/) before. Even though it has a lot of interesting features, I liked the nginx based simplicity of the proxy proposed by the project. So even though the recipe needed a bit of tweaking, I'm glad with this help I got it working.

Thanks again!

**Reply by tommoor** [reply]
✅  Added these lines to the documentation
